servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java

```diff
@@ -0,0 +1,721 @@
+    }
+
+    @Override
+    public Publisher<ServiceDiscovererEvent<InetSocketAddress>> dnsSrvQuery(final String serviceName) {
+        return new SrvRecordPublisher(serviceName).flatMapConcatIterable(identity())
+                .<ServiceDiscovererEvent<InetSocketAddress>>flatMapMergeSingle(srvEvent -> {
+            assertInEventloop();
+
+            if (srvEvent.isAvailable()) {
+                ARecordPublisher entry = new ARecordPublisher(srvEvent.address().hostName()) {
+                    @Override
+                    protected void onFinallyCleanup() {
+                        removeSrvEntry0(this);
+                    }
+                };
+                addSrvEntry0(entry);
+                return entry.flatMapConcatIterable(identity()).map(inetEvent -> new DefaultServiceDiscovererEvent<>(
+                        new InetSocketAddress(inetEvent.address(), srvEvent.address().port()), inetEvent.isAvailable())
+                       ).firstOrElse(() -> {
+                           LOGGER.info("No A/AAAA records found for hostname {} corresponding to SRV name {}",
+                                   srvEvent.address().hostName(), serviceName);
+                           return null;
+                       });
+            } else {
+                final List<ARecordPublisher> inetWatchers = srvARecordMap.remove(srvEvent.address().hostName());
+                if (inetWatchers != null) {
+                    for (ARecordPublisher inetWatcher : inetWatchers) {
+                        inetWatcher.close0();
+                    }
+                }
+
+                // We will filter the null value out on the outer flatMap transformation.
+                return Single.succeeded(null);
+            }
+        })
+        // When the SRV record becomes unavailable we clean up state and return a Single.succeeded(null), but
+        // we don't want to propagate this downstream, so filter it out.
+        .filter(Objects::nonNull);
+    }
+
+    private void addSrvEntry0(final ARecordPublisher entry) {
+        assertInEventloop();
+
+        srvARecordMap.computeIfAbsent(entry.inetHost, k -> new ArrayList<>(2)).add(entry);
+    }
+
+    private void removeSrvEntry0(final ARecordPublisher entry) {
+        assertInEventloop();
+
+        final List<ARecordPublisher> inetWatchers = srvARecordMap.get(entry.inetHost);
+        if (inetWatchers == null) {
+            return;
+        }
+        inetWatchers.remove(entry);
+        if (inetWatchers.isEmpty()) {
+            srvARecordMap.remove(entry.inetHost);
+        }
+    }
+
+    @Override
+    public Completable onClose() {
+        return asyncCloseable.onClose();
+    }
+
+    @Override
+    public Completable closeAsync() {
+        return asyncCloseable.closeAsync();
+    }
+
+    @Override
+    public Completable closeAsyncGracefully() {
+        return asyncCloseable.closeAsyncGracefully();
+    }
+
+    private void closeAsync0() {
```

Comment:
So previously[1] we were terminating all active queries in `close()`. I don't see a test for the close behavior in the current `master` but looks like this something that is missing from this change?

[1] https://github.com/apple/servicetalk/blob/master/servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscoverer.java#L209-L223

Reply:
> Remove tracking of all outstanding streams and close() cleanup to
reduce state. Subscribers of these sources (e.g. LoadBalancers) are
responsible for cancelling if they are no longer interested in demand.
There is best effort close cleanup when TTLs timeout fire in the event
the DnsClient is closed but the streams are not cancelled (which is not
an expected use case).

This was removed as it isn't guaranteed by the interface and subscribers are currently responsible for managing lifetime/cancellation. Removing this constraint simplified the code but we can add it back if necessary.